File: ws_pty_verify.js
const wsUrl = "ws://127.0.0.1:8080/ws?token=test";
const cmd = "printf \"Enter your name: \"; read name; printf \"Hello, %s\\n\" \"$name\"";

if (typeof WebSocket === "undefined") {
  console.error("当前 Node.js 没有内置 WebSocket。请使用较新的 Node.js 版本运行此脚本。");
  process.exit(1);
}

const ws = new WebSocket(wsUrl);
let inputSent = false;
let sawHello = false;

function sendExec() {
  const payload = { action: "exec", mode: "pty", cmd };
  console.log(">>> SEND", JSON.stringify(payload));
  ws.send(JSON.stringify(payload));
}

function sendInput() {
  if (inputSent) return;
  inputSent = true;
  const payload = { action: "input", data: "VibeCoder\n" };
  console.log(">>> SEND", JSON.stringify(payload));
  ws.send(JSON.stringify(payload));
}

function shouldAutoInput(event) {
  if (!event || typeof event !== "object") return false;
  if (event.type === "prompt_request") return true;
  if (event.type === "log" && typeof event.msg === "string") {
    const text = event.msg.toLowerCase();
    return text.includes("enter your name") || text.includes("name:") || text.endsWith(":") || text.endsWith(": ");
  }
  return false;
}

ws.addEventListener("open", () => {
  console.log("WebSocket 已连接:", wsUrl);
  sendExec();
});

ws.addEventListener("message", (evt) => {
  const raw = typeof evt.data === "string" ? evt.data : String(evt.data);
  console.log("<<< RECV", raw);

  let event;
  try {
    event = JSON.parse(raw);
  } catch {
    return;
  }

  if (shouldAutoInput(event)) {
    sendInput();
  }

  if (event.type === "log" && typeof event.msg === "string" && event.msg.includes("Hello, VibeCoder")) {
    sawHello = true;
    console.log("✅ 已收到目标输出: Hello, VibeCoder");
  }

  if (event.type === "error") {
    console.error("❌ 收到错误事件:", event.msg || event);
  }

  if (event.type === "session_state" && event.state === "closed") {
    console.log("会话结束:", event.msg || "");
    process.exit(sawHello ? 0 : 2);
  }
});

ws.addEventListener("error", (err) => {
  console.error("WebSocket 错误:", err);
  process.exit(4);
});

setTimeout(() => {
  if (!sawHello) {
    console.error("❌ 超时：未在预期时间内收到 Hello, VibeCoder");
    try {
      ws.close();
    } catch {}
    process.exit(3);
  }
}, 30000);

File: ws_pty_verify.py
#!/usr/bin/env python3
import asyncio
import json
import sys

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
if hasattr(sys.stderr, "reconfigure"):
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

WS_URL = "ws://127.0.0.1:8080/ws?token=test"
COMMAND = r'''python3 -c "name = input('Enter your name: '); print('Hello, ' + name)"'''

try:
    import websockets
except ImportError:
    print("缺少 websockets 库。请先安装：pip install websockets", file=sys.stderr)
    sys.exit(1)


def should_auto_input(event: dict) -> bool:
    if not isinstance(event, dict):
        return False

    if event.get("type") == "prompt_request":
        return True

    if event.get("type") == "log":
        msg = event.get("msg")
        if isinstance(msg, str):
            text = msg.lower()
            return (
                "enter your name" in text
                or "name:" in text
                or text.endswith(":")
                or text.endswith(": ")
            )

    return False


async def main():
    input_sent = False
    saw_hello = False

    async with websockets.connect(WS_URL) as ws:
        exec_payload = {
            "action": "exec",
            "mode": "pty",
            "cmd": COMMAND,
        }
        print(">>> SEND", json.dumps(exec_payload, ensure_ascii=False))
        await ws.send(json.dumps(exec_payload))

        while True:
            raw = await asyncio.wait_for(ws.recv(), timeout=30)
            print("<<< RECV", raw)

            try:
                event = json.loads(raw)
            except json.JSONDecodeError:
                continue

            if should_auto_input(event) and not input_sent:
                input_sent = True
                input_payload = {
                    "action": "input",
                    "data": "VibeCoder\n",
                }
                print(">>> SEND", json.dumps(input_payload, ensure_ascii=False))
                await ws.send(json.dumps(input_payload))

            if event.get("type") == "log" and "Hello, VibeCoder" in str(event.get("msg", "")):
                saw_hello = True
                print("已收到目标输出: Hello, VibeCoder")

            if event.get("type") == "error":
                print(f"收到错误事件: {event.get('msg')}", file=sys.stderr)

            if event.get("type") == "session_state" and event.get("state") == "closed":
                print("会话结束:", event.get("msg", ""))
                sys.exit(0 if saw_hello else 2)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except asyncio.TimeoutError:
        print("超时：未在预期时间内收到 Hello, VibeCoder", file=sys.stderr)
        sys.exit(3)
    except KeyboardInterrupt:
        print("已取消")
        sys.exit(130)


def should_auto_input(event: dict) -> bool:
    if not isinstance(event, dict):
        return False

    if event.get("type") == "prompt_request":
        return True

    if event.get("type") == "log":
        msg = event.get("msg")
        if isinstance(msg, str):
            text = msg.lower()
            return (
                "enter your name" in text
                or "name:" in text
                or text.endswith(":")
                or text.endswith(": ")
            )

    return False


async def main():
    input_sent = False
    saw_hello = False

    async with websockets.connect(WS_URL) as ws:
        exec_payload = {
            "action": "exec",
            "mode": "pty",
            "cmd": COMMAND,
        }
        print(">>> SEND", json.dumps(exec_payload, ensure_ascii=False))
        await ws.send(json.dumps(exec_payload))

        while True:
            raw = await asyncio.wait_for(ws.recv(), timeout=30)
            print("<<< RECV", raw)

            try:
                event = json.loads(raw)
            except json.JSONDecodeError:
                continue

            if should_auto_input(event) and not input_sent:
                input_sent = True
                input_payload = {
                    "action": "input",
                    "data": "VibeCoder\n",
                }
                print(">>> SEND", json.dumps(input_payload, ensure_ascii=False))
                await ws.send(json.dumps(input_payload))

            if event.get("type") == "log" and "Hello, VibeCoder" in str(event.get("msg", "")):
                saw_hello = True
                print("已收到目标输出: Hello, VibeCoder")

            if event.get("type") == "error":
                print(f"收到错误事件: {event.get('msg')}", file=sys.stderr)

            if event.get("type") == "session_state" and event.get("state") == "closed":
                print("会话结束:", event.get("msg", ""))
                sys.exit(0 if saw_hello else 2)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except asyncio.TimeoutError:
        print("超时：未在预期时间内收到 Hello, VibeCoder", file=sys.stderr)
        sys.exit(3)
    except KeyboardInterrupt:
        print("已取消")
        sys.exit(130)
